library/alloc/src/collections/vec_deque/drain.rs

```diff
@@ -57,31 +57,17 @@ impl<'a, T, A: Allocator> Drain<'a, T, A> {
     unsafe fn as_slices(&self) -> (*mut [T], *mut [T]) {
         unsafe {
             let deque = self.deque.as_ref();
-            // FIXME: This is doing almost exactly the same thing as the else branch in `VecDeque::slice_ranges`.
-            // Unfortunately, we can't just call `slice_ranges` here, as the deque's `len` is currently
-            // just `drain_start`, so the range check would (almost) always panic. Between temporarily
-            // adjusting the deques `len` to call `slice_ranges`, and just copy pasting the `slice_ranges`
-            // implementation, this seemed like the less hacky solution, though it might be good to
-            // find a better one in the future.
-
-            // because `self.remaining != 0`, we know that `self.idx < deque.original_len`, so it's a valid
-            // logical index.
-            let wrapped_start = deque.to_physical_idx(self.idx);
-
-            let head_len = deque.capacity() - wrapped_start;
-
-            let (a_range, b_range) = if head_len >= self.remaining {
-                (wrapped_start..wrapped_start + self.remaining, 0..0)
-            } else {
-                let tail_len = self.remaining - head_len;
-                (wrapped_start..deque.capacity(), 0..tail_len)
-            };
-
-            // SAFETY: the range `self.idx..self.idx+self.remaining` lies strictly inside
-            // the range `0..deque.original_len`. because of this, and because of the fact
-            // that we acquire `a_range` and `b_range` exactly like `slice_ranges` would,
+
+            let start = self.idx;
+            // We know that `self.idx + self.remaining <= deque.len <= usize::MAX`, so this won't overflow.
+            let end = start + self.remaining;
+
+            // SAFETY: the range `start..end` lies strictly inside
+            // the range `0..deque.original_len`. Because of this, and because
```

Comment:
Hrm, this means that end may be smaller than original_len. That happens because one is draining from the middle of the deque, right? But then end isn't deque.len anymore which one is supposed to pass to `slice_ranges` (according to the new documentation).

Reply:
In this case it doesn't really matter which length you pass, as long as it's not smaller than `end`. Since the range is trivially in bounds, and we don't pass a `RangeFrom` to `slice_ranges` (which would then use the `len` parameter as the end index), the bounds checking in `slice_ranges` is technically redundant anyways. Explicitly calculating `original_len` here to pass as a dead bounds check would just result in slightly more unnecessary computations with no difference in behavior.

Reply:
Well, the documentation should say so? Also, could passing in the wrong length cause any unsafety?

Reply:
The function itself can't cause any unsafety, it doesn't access any memory and only computes index ranges into the physical buffer. Using it with the wrong inputs and then using the return values to index into the buffer could obviously lead to buffer overruns, for example:
```rs
let (a, b) = self.slice_ranges(.., usize::MAX);
// unsafety is introduced here since `b` doesn't satisfy the `buffer_range` preconditions.
let b = unsafe { &*self.buffer_range(b) };
```
I'm not 100% sure how to word the safety docs for this, as you can't just require the first `len` items to be initialized (because `Drain` may have already destroyed elements at the beginning of the deque). Would something like
 "The caller has to ensure that for all possible inputs, the result of calling `slice::range(range, ..len)` is a valid range into the logical array, and all of its elements must be initialized"
be okay? It's kind of a mouthful, but I'm not sure how else to phrase it.